cns/kubecontroller/nodenetworkconfig/conversion.go

```diff
@@ -87,13 +91,41 @@ func CreateNCRequestFromStaticNC(nc v1alpha.NetworkContainer) (*cns.CreateNetwor
 
 	secondaryIPConfigs := map[string]cns.SecondaryIPConfig{}
 
-	// iterate through all IP addresses in the subnet described by primaryPrefix and
-	// add them to the request as secondary IPConfigs.
-	for addr := primaryPrefix.Masked().Addr(); primaryPrefix.Contains(addr); addr = addr.Next() {
-		secondaryIPConfigs[addr.String()] = cns.SecondaryIPConfig{
-			IPAddress: addr.String(),
-			NCVersion: int(nc.Version),
+	switch nc.Type {
+	case v1alpha.VNETBlock:
+		{
+			for blockIndex, ipAssignment := range nc.IPAssignments {
+				cidrPrefix, err := netip.ParsePrefix(ipAssignment.IP)
+				if err != nil {
+					return nil, errors.Wrapf(ErrInvalidCidrIPBlock, "CIDR IP Block: %s", ipAssignment.IP)
+				}
+
+				zeroAddr := cidrPrefix.Masked().Addr()
+				if blockIndex == 0 {
+					zeroAddr = zeroAddr.Next() // # Block 1 CIDR. Exclude 1st IP from pod assignment
+				}
```

Comment:
@nairashu 
I am making an assumption on intent here, but I read this as blocking the first address of the first Secondary IP Assignment Block for use as the NC Primary IP. 

If that is correct:
I'm concerned that we're blocking the Primary IP from the "first" Block in the NC Secondary IP Assignments with an expectation of consistent ordering when that is really not guaranteed.

Without changing too much, maybe we could make the "PrimaryIP" a CIDR block like any of the Secondary IP Assignments, and use the 0th address of that block as the Primary IP for the NC while adding the rest of the block as pod-assignable IPs, as if they were Secondary assignments (this is what we do in Overlay). Then, any additional blocks that we receive through the Secondary IP Assignments do not need special-case handling like this.

Reply:
Had an offline discussion with Evan on this front. Few clarifying points:

1. Vnet Scale is a static allocation for the node as there will be no scaling, so all the blocks are assigned during node creation.
2. Evan brought up the concern that slices intrinsically do not have ordering, especially since we are converting them from JSON back and forth.
3. Will have to change the NNC where the primary IP CIDR block is not included in the IP asssignments section when being sent.

@aggarwal0009 : Change logic to:

1. We will need to process the primary IP CIDR block from the primary IP section where the 0th IP is excluded and the remaining IPs are added to the secondary IP slice.
2. All assignments in IP assignments can now be added directly to the secondary IP slice after flattening.

Reply:
Updated